src/traversal.rs

```diff
@@ -0,0 +1,134 @@
+                let member = aar.get_mut(index).ok_or(anyhow!("Array out of range"))?;
+                do_traverse(path, &mut At::Table(member), op)
+            }
+            At::Item(item) => match item {
+                Item::ArrayOfTables(aar) => At::ArrayOfTables(aar).fold(key, path, op),
+                Item::Table(table) => At::Table(table).fold(key, path, op),
+                Item::Value(value) => At::Value(value).fold(key, path, op),
+                _ => Err(anyhow!("Unable to index item {:?} with {:?}", item, key)),
+            },
+            At::Table(table) => {
+                let mut found = table
+                    .get_mut(key)
+                    .ok_or(anyhow!("Unable to index table with {:?}", key))?;
+                do_traverse(path, &mut At::Item(&mut found), op)
+            }
+            At::Value(value) => match value {
+                Value::Array(arr) => At::Array(arr).fold(key, path, op),
+                _ => Err(anyhow!("Unable to index value {:?} with {:?}", value, key)),
+            },
+        }
+    }
+
+    pub fn to_value(&mut self) -> Result<serde_json::Value> {
+        match self {
+            At::Array(_arr) => Err(anyhow!("array")),
+            At::ArrayOfTables(_aar) => Err(anyhow!("stub")),
+            At::Item(item) => match item {
+                Item::None => Ok(Json::Null),
+                Item::Value(value) => At::Value(value).to_value(),
+                Item::ArrayOfTables(aar) => At::ArrayOfTables(aar).to_value(),
+                Item::Table(table) => At::Table(table).to_value(),
+            },
+            At::Value(value) => match value {
+                Value::String(s) => {
+                    s.fmt();
+                    Ok(Json::String(s.value().clone()))
+                }
+                Value::Integer(i) => Ok(Json::Number(serde_json::Number::from(
+                    i.clone().into_value(),
+                ))),
+                Value::Float(f) => {
+                    let n = serde_json::Number::from_f64(f.clone().into_value())
+                        .expect("float infinite and nan not allowed");
+                    Ok(Json::Number(n))
+                }
+                Value::Boolean(b) => Ok(Json::Bool(b.clone().into_value())),
+                Value::Array(arr) => {
+                    let xs = arr
+                        .iter_mut()
+                        .map(|val| At::Value(val).to_value())
+                        .collect::<Result<Vec<Json>>>()?;
+                    Ok(Json::Array(xs))
+                }
+                Value::Datetime(dt) => Ok(Json::String(dt.to_string())),
+                Value::InlineTable(table) => {
+                    let inner: Map<String, Json> = table
+                        .iter_mut()
+                        .map(|(k, v)| At::Value(v).to_value().map(|v| (k.to_string(), v)))
+                        .collect::<Result<Map<String, Json>>>()?;
+                    Ok(Json::Object(inner))
```

Comment:
not-even-a-nit but you can remove some type annotations
```suggestion
                    let inner = table
                        .iter_mut()
                        .map(|(k, v)| At::Value(v).to_value().map(|v| (k.to_string(), v)))
                        .collect::<Result<_>>()?;
                    Ok(Json::Object(inner))
```

Reply:
Accepted!